### ISSUE TITLE
scipy notebook wth IOOS installed is 9+ GB

### ISSUE THREAD

**Issue by isuftin** [USGS-CMG/data-life-cycle-cloud-docker-jupyterhub#9]
The image is too large to be very usable in a cluster due to loading times and network transfer from datacenter to AWS. 

Perform cleanup tasks during layer creation in Docker image for a final image that is smaller.

**Comment by rsignell-usgs** [comment]
@ocefpaf, just for grins...

**Comment by ocefpaf** [comment]
@isuftin are you referring to the IOOS conda environment?
If so, bare in mind that it was not designed for such use case.
The environment was designed to help users have a stand-alone (matlab-like) work enviroment.
If you do need to use some of the software available there I recommend you to create your own env based on our [file](https://github.com/ioos/notebooks_demos/blob/master/environment.yml), where you can remove everything that you won't be using.

**Comment by isuftin** [comment]
@ocefpaf Consider that I know next to nothing about Jupyterhub or the actual use case @rsignell-usgs is envisioning. So right now I am able to get the Docker image down to about ~8.5G by issuing a `conda clean --all` after each install command. 

@rsignell-usgs Any insights into what should be pared down?

**Comment by ocefpaf** [comment]
Ah OK. I guess that @rsignell-usgs may want to preserve as much as possible from that env then.

I imagine you can skip the lines in https://github.com/USGS-CMG/data-life-cycle-cloud-docker-jupyterhub/blob/564d89026ca1bb6b1b37437992222c500b817a26/files/jupyter-docker-stacks/single-user/Dockerfile

```
RUN conda update --yes --all
RUN conda install anaconda-navigator
```

B/c the env will have updated packages you don't need to update the root env.
Also, I don't think the navigator is needed on *nix machines. (That is mostly for Windows.)

I'll take a look at the env and sort the packages by size to help you guys decide where to make cuts.

**Comment by isuftin** [comment]
@ocefpaf Take a peek at the latest PR on this project to see the cleanup I've done. I can try to work with your suggestions tomorrow to see if it works still. 

I was trying to follow http://ioos.github.io/notebooks_demos/other_resources/ which did not specify that navigator is a windows-only setup.

**Comment by isuftin** [comment]
@ocefpaf You can also take a look @ https://cida.usgs.gov/artifactory/webapp/#/artifacts/browse/simple/DockerV2Info/owi-docker-development/dlcc/jupyter-singleuser/latest to see what's taking up the space currently

**Comment by ocefpaf** [comment]
> I was trying to follow http://ioos.github.io/notebooks_demos/other_resources/ which did not specify that navigator is a windows-only setup.

Yep. I know. That is why I comment on those. The instructions is target to users and not a service like that. (Only users will need those two steps.)

> @ocefpaf Take a peek at the latest PR on this project to see the cleanup I've done.

LGTM. I believe you can call the `conda clean` only once at the end.

> @ocefpaf You can also take a look @ https://cida.usgs.gov/artifactory/webapp/#/artifacts/browse/simple/DockerV2Info/owi-docker-development/dlcc/jupyter-singleuser/latest to see what's taking up the space currently

Nice viz! I'll explore it and get back to you.

**Comment by isuftin** [comment]
> LGTM. I believe you can call the conda clean only once at the end.

@ocefpaf Unfortunately doing that once at the very end still crates Docker image layers of the bloated size without cleanup. You must include the cleanup step along with the install step in the same RUN line so that those files that get cleaned up don't persist to the next step.

**Comment by ocefpaf** [comment]
> You must include the cleanup step along with the install step in the same RUN line so that those files that get cleaned up don't persist to the next step.

indeed. I forgot how docker stack things.

**Comment by isuftin** [comment]
With the conda clean commands in the Dockerfile, the final compressed image in Artifactory went from 3.5GB to 2.7GB. Not bad. 

Will try to clean further tomorrow.

**Comment by isuftin** [comment]
@rsignell-usgs in regards to https://github.com/USGS-CMG/data-life-cycle-cloud-docker-jupyterhub/issues/9#issuecomment-324140415 -- If you take a look @ https://github.com/ioos/notebooks_demos/blob/master/environment.yml do you know if any of these packages are not necessary?

**Comment by ocefpaf** [comment]
Here is a list of packages by size to help you with that task:

https://gist.github.com/ocefpaf/da69c4374d37bf60b84d83228fb8ae30

I would start with the all the R packages, `cython`, `pandoc`, `jupyter_contrib_nbextensions`, `sympy`, `spyder`, `botocore`, and `pysal`. Those are quite big and rarely used. If `gdal` is not needed I recommend removing that too. There are many dependencies up-/down-stream of `gdal` that are quite big.

**Comment by isuftin** [comment]
With the latest update the image is down to 2.5G -- https://github.com/USGS-CMG/data-life-cycle-cloud-docker-jupyterhub/compare/master@%7B1day%7D...master#diff-7e72e4037408c1f3719206c3ff5c9c14

I've still yet to implement your latest comment @ocefpaf - I want to wait on @rsignell-usgs to acknowledge the changes you suggest so I don't spend time going back and forth.

**Comment by rsignell-usgs** [comment]
If we are okay with the 2.5G image, let's leave it the way it is.   It would be nice to have the same environment as the many other folks we've pointed to the IOOS environment and we might need those tools.

**Comment by isuftin** [comment]
@rsignell-usgs The IOOS environment should be functional at this point. Because of the the issue you see here ( https://github.com/USGS-CMG/data-life-cycle-cloud-docker-jupyterhub/issues/5#issuecomment-324451930 ), the Docker notebook image file size will probably have to come back up until Conda 4.4 is out (and hopefully they fix this issue there)

**Comment by rsignell-usgs** [comment]
Us modelers are not big on R, but I know a lot of folks in the USGS community use R.   If it's important to cut the size down, let's remove them.   We can always add them back in later.

**Comment by isuftin** [comment]
I'll test around with removing R and see what the resulting image size becomes. Ultimately, it's going to be a balance between image size and usability and I feel (as do you, probably) that usability has a much higher priority.

**Comment by isuftin** [comment]
@rsignell-usgs I've tested the Docker build without R. At best it saves 100MB so let's not worry about it. Let's consider the current compressed image size of 2.4GB to be our base size and go from there.
